File: chicky/src/controllers/lightController.js
/**
 * Light controller for handling light-related commands
 */
const logger = require('../utils/logger');
const { isWithinAllowedHours } = require('../utils/validation');
const { controlLight } = require('../hardware/hardwareInterface');

/**
 * Handle light command from remote server
 * @param {Object} data - Command data
 * @param {Object} config - Configuration object
 * @param {Object} socket - Socket.IO client
 */
function handleLightCommand(data, config, socket) {
  try {
    // Check if the command is within allowed hours
    if (!isWithinAllowedHours(config)) {
      // If configuration is not available, provide a generic message
      const message = config.allowed_start_hour === null || config.allowed_end_hour === null
        ? 'Light command rejected: waiting for configuration from server'
        : `Light command rejected: outside allowed hours (${config.allowed_start_hour}am-${config.allowed_end_hour > 12 ? (config.allowed_end_hour - 12) + 'pm' : config.allowed_end_hour + 'am'})`;
      
      logger.warn(message);
      
      socket.emit('light-confirmation', {
        success: false,
        state: data.state,
        error: config.allowed_start_hour === null || config.allowed_end_hour === null
          ? 'Sorry, the system is still initializing. Please try again in a moment.'
          : `Sorry, the chicken coop light can only be operated between ${config.allowed_start_hour}am and ${config.allowed_end_hour > 12 ? (config.allowed_end_hour - 12) + 'pm' : config.allowed_end_hour + 'am'}.`
      });
      return;
    }

    // Control the light
    const result = controlLight(data.state);
    
    // Send confirmation to the server
    socket.emit('light-confirmation', result);
  } catch (error) {
    logger.error('Error handling light command:', error);
    socket.emit('light-confirmation', {
      success: false,
      state: data.state,
      error: error.message
    });
  }
}

/**
 * Automatically turn off lights at the configured end hour
 * @param {Object} config - Configuration object
 * @param {Object} socket - Socket.IO client
 * @param {boolean} connected - Whether the socket is connected
 */
function autoShutoffLight(config, socket, connected) {
  // Check if configuration is available
  if (config.allowed_end_hour === null) {
    logger.warn('Auto-shutdown skipped: configuration not yet available');
    return;
  }
  
  logger.info(`Auto-shutdown: Turning off lights at configured end hour (${config.allowed_end_hour})`);
  
  try {
    // Turn off the light
    const result = controlLight('off');
    
    // Notify the server about the state change if connected
    if (connected && result.success) {
      socket.emit('light-confirmation', {
        success: true,
        state: 'off',
        automatic: true
      });
    }
  } catch (error) {
    logger.error('Error in automatic light shutdown:', error);
  }
}

module.exports = {
  handleLightCommand,
  autoShutoffLight
}; 

File: chicky/src/controllers/light_controller.py
"""
Light controller for handling light-related commands
"""
from typing import Dict, Any, Callable, Optional
import asyncio

from src.utils.logger import get_logger
from src.utils.validation import is_within_allowed_hours
from src.hardware.hardware_interface import control_light

logger = get_logger(__name__)

async def handle_light_command(data: Dict[str, Any], config: Dict[str, Any], socket) -> Dict[str, Any]:
    """
    Handle light command from remote server
    """
    try:
        # Check if the command is within allowed hours
        if not is_within_allowed_hours(config):
            # If configuration is not available, provide a generic message
            if config.get("allowed_start_hour") is None or config.get("allowed_end_hour") is None:
                message = "Light command rejected: waiting for configuration from server"
                error_msg = "Sorry, the system is still initializing. Please try again in a moment."
            else:
                allowed_start = config["allowed_start_hour"]
                allowed_end = config["allowed_end_hour"]
                end_display = f"{allowed_end - 12}pm" if allowed_end > 12 else f"{allowed_end}am"
                message = f"Light command rejected: outside allowed hours ({allowed_start}am-{end_display})"
                error_msg = f"Sorry, the chicken coop light can only be operated between {allowed_start}am and {end_display}."
            
            logger.warning(message)
            
            return {
                "success": False,
                "state": data.get("state", "unknown"),
                "error": error_msg
            }

        # Control the light
        result = control_light(data.get("state", "off"))
        
        # Emit the result to the server
        if socket and hasattr(socket, "emit"):
            try:
                await socket.emit("light_result", result)
            except Exception as e:
                logger.error(f"Error emitting light result: {e}")
        
        return result
    except Exception as e:
        logger.error(f"Error handling light command: {e}")
        return {
            "success": False,
            "state": data.get("state", "unknown"),
            "error": str(e)
        }

async def auto_shutoff_light(socket, is_connected: Callable[[], bool]) -> Dict[str, Any]:
    """
    Automatically shut off the light at the end of the allowed hours
    """
    try:
        logger.info("Auto shutoff: turning light off")
        
        # Control the light
        result = control_light("off")
        
        # Emit the result to the server if connected
        if is_connected() and socket and hasattr(socket, "emit"):
            try:
                await socket.emit("light_result", {
                    "success": result["success"],
                    "state": "off",
                    "auto": True
                })
            except Exception as e:
                logger.error(f"Error emitting auto shutoff result: {e}")
        
        return result
    except Exception as e:
        logger.error(f"Error in auto shutoff: {e}")
        return {
            "success": False,
            "error": str(e)
        }